File: medium/Hash-Table/group-anagrams/group_anagrams.py
from typing import List

def groupAnagrams(strs: List[str]) -> List[List[str]]:
    """
    Group anagrams together from a list of strings
    Args:
        strs: List[str] - Input list of strings
    Returns:
        List[List[str]] - Groups of anagrams
    Time Complexity: O(n * k log k) - n = number of strings, k = max length of a string
    Space Complexity: O(n * k) - Store all strings in the hash map
    """
    anagram_map = {}

    for s in strs:
        # Sort the string to get the key
        key = ''.join(sorted(s))

        # Add current string to the corresponding list
        if key not in anagram_map:
            anagram_map[key] = []
        anagram_map[key].append(s)

    # Return all groups as a list
    return list(anagram_map.values())

# Test
if __name__ == "__main__":
    test = ["eat","tea","tan","ate","nat","bat"]
    print(groupAnagrams(test))

File: medium/Hash-Table/group-anagrams/group_anagrams.cpp
#include <vector>
#include <string>
#include <unordered_map>
#include <algorithm>
using namespace std;

/**
 * Group anagrams together from a list of strings
 * @param strs: vector<string> - Input list of strings
 * @return: vector<vector<string>> - Groups of anagrams
 * Time Complexity: O(n * k log k) - n = number of strings, k = max length of a string
 * Space Complexity: O(n * k) - Store all strings in the hash map
 */
vector<vector<string>> groupAnagrams(vector<string>& strs) {
    unordered_map<string, vector<string>> anagram_map;

    for (string& s : strs) {
        string key = s;
        sort(key.begin(), key.end());
        anagram_map[key].push_back(s);
    }

    vector<vector<string>> result;
    for (auto& pair : anagram_map) {
        result.push_back(pair.second);
    }

    return result;
}

// Test
#include <iostream>
int main() {
    vector<string> test = {"eat","tea","tan","ate","nat","bat"};
    auto res = groupAnagrams(test);
    for (auto& g : res) {
        for (string& s : g) cout << s << " ";
        cout << endl;
    }
    return 0;
}
